### PR TITLE
console: allow changing owners of secrets

### PR DESCRIPTION
**Pull request by d0s056f**

### CODE REVIEW

File: server/impl/src/main/java/com/walmartlabs/concord/server/org/secret/SecretManager.java
Comment (ibodrov) [inline server/impl/src/main/java/com/walmartlabs/concord/server/org/secret/SecretManager.java:359]: Other resources require `OWNER` privileges to change the owner. See `ProjectManager#update` for example.
I think we should follow the same logic here.

File: server/impl/src/main/java/com/walmartlabs/concord/server/org/secret/SecretManager.java
Comment (ibodrov) [inline server/impl/src/main/java/com/walmartlabs/concord/server/org/secret/SecretManager.java:87]: Unused?

File: server/impl/src/main/java/com/walmartlabs/concord/server/org/secret/SecretManager.java
Comment (d0s056f) [reply server/impl/src/main/java/com/walmartlabs/concord/server/org/secret/SecretManager.java:359]: I couldn't update the secret when I was signed in as another user while testing. could only do it when signed in as the owner

Comment (ibodrov) [reply server/impl/src/main/java/com/walmartlabs/concord/server/org/secret/SecretManager.java:358]: That's the point.

Comment (d0s056f) [reply server/impl/src/main/java/com/walmartlabs/concord/server/org/secret/SecretManager.java:354]: right. I meant we are already following the right logic, no?
what do I need to change then

Comment (ibodrov) [reply server/impl/src/main/java/com/walmartlabs/concord/server/org/secret/SecretManager.java:354]: Interesting. Where's the check for that then? Similar to https://github.com/walmartlabs/concord/blob/master/server/impl/src/main/java/com/walmartlabs/concord/server/org/project/ProjectManager.java#L142-L151

The current code checks only for `WRITER` permissions https://github.com/walmartlabs/concord/blob/master/server/impl/src/main/java/com/walmartlabs/concord/server/org/secret/SecretManager.java#L265-L271

**Comment by ibodrov** [comment]
An integration test would be nice :-)

File: server/impl/src/main/java/com/walmartlabs/concord/server/org/secret/SecretManager.java
Comment (d0s056f) [reply server/impl/src/main/java/com/walmartlabs/concord/server/org/secret/SecretManager.java:354]: there is no explicit check here because 

> An integration test would be nice :-)

yep will do

File: server/impl/src/main/java/com/walmartlabs/concord/server/org/secret/SecretUpdateRequest.java
Comment (ibodrov) [inline server/impl/src/main/java/com/walmartlabs/concord/server/org/secret/SecretUpdateRequest.java:48]: Also, as @brig pointed out, other entities use `Owner` structure. I think we should use it here too, for symmetry.

File: server/impl/src/main/java/com/walmartlabs/concord/server/org/secret/SecretManager.java
Comment (d0s056f) [reply server/impl/src/main/java/com/walmartlabs/concord/server/org/secret/SecretManager.java]: ok so I added the check for secret#update#ownerId, but I still think the access level for that should be WRITER. 3 reasons - 
1. all other secret update options require only WRITER access, including delete method (which i changed to owner in this PR)
2. none of the secret#update actions, including ownerId change, can really harm anyone's concord setup. One can always recreate a secret as long as they have the password. if they don't, what are they doing as owners? :)
3. if the owner leaves the company, team members with write access should be able to change it without bothering us admins.
Secret#update is less critical than org and project in that sense. It's obviously your call in the end.. so let me know and i'll change it to OWNER if so

Comment (d0s056f) [reply server/impl/src/main/java/com/walmartlabs/concord/server/org/secret/SecretManager.java]: @ibodrov

File: server/impl/src/main/java/com/walmartlabs/concord/server/org/secret/SecretUpdateRequest.java
Comment (d0s056f) [reply server/impl/src/main/java/com/walmartlabs/concord/server/org/secret/SecretUpdateRequest.java:48]: missed this. I'm on it..

File: server/impl/src/main/java/com/walmartlabs/concord/server/org/secret/SecretManager.java
Comment (ibodrov) [reply server/impl/src/main/java/com/walmartlabs/concord/server/org/secret/SecretManager.java]: I personally prefer symmetry with other kind of entities over the concerns above.

1. yes, the `DELETE` operation should require `OWNER` privileges;
2. sure, but that's where symmetry and consistency with other types of entities is better than a special case;
3. multiple people can be OWNERs including the org's OWNER.

Let's vote? @walmartlabs/concord :+1: this comment for requiring OWNER privileges, :+1: @d0s056f's post for lowering it to `WRITER`

Comment (brig) [reply server/impl/src/main/java/com/walmartlabs/concord/server/org/secret/SecretManager.java]: 👍 for symmetry :)

Comment (d0s056f) [reply server/impl/src/main/java/com/walmartlabs/concord/server/org/secret/SecretManager.java]: haha I'm convinced.. will do :)

**Comment by d0s056f** [comment]
@ibodrov @brig

**Comment by ibodrov** [comment]
@concord-ci test it

**Comment by ibodrov** [comment]
@d0s056f some tests are failing
```
10:45:44 [INFO ] [ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 1.549 s <<< FAILURE! - in com.walmartlabs.concord.it.server.SecretIT
10:45:44 [INFO ] [ERROR] testPublicKey(com.walmartlabs.concord.it.server.SecretIT)  Time elapsed: 1.549 s  <<< ERROR!
10:45:44 [INFO ] com.walmartlabs.concord.ApiException: User not found: 4b9d496a-c3a0-4e1b-804c-ac3fccddcb27
10:45:44 [INFO ] 	at com.walmartlabs.concord.it.server.SecretIT.testPublicKey(SecretIT.java:59)
```

```
10:38:21 [INFO ] [ERROR] Tests run: 12, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 3.693 s <<< FAILURE! - in com.walmartlabs.concord.it.server.TeamRbacIT
10:38:21 [INFO ] [ERROR] testSecretAccessLevels(com.walmartlabs.concord.it.server.TeamRbacIT)  Time elapsed: 0.33 s  <<< ERROR!
10:38:21 [INFO ] com.walmartlabs.concord.ApiException: Server response: The current user doesn't have the necessary access level (OWNER) to the secret: secret_1599057498237_7a008a
10:38:21 [INFO ] 	at com.walmartlabs.concord.it.server.TeamRbacIT.testSecretAccessLevels(TeamRbacIT.java:728)
```

**Comment by d0s056f** [comment]
> @d0s056f some tests are failing
> 
> ```
> 10:45:44 [INFO ] [ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 1.549 s <<< FAILURE! - in com.walmartlabs.concord.it.server.SecretIT
> 10:45:44 [INFO ] [ERROR] testPublicKey(com.walmartlabs.concord.it.server.SecretIT)  Time elapsed: 1.549 s  <<< ERROR!
> 10:45:44 [INFO ] com.walmartlabs.concord.ApiException: User not found: 4b9d496a-c3a0-4e1b-804c-ac3fccddcb27
> 10:45:44 [INFO ] 	at com.walmartlabs.concord.it.server.SecretIT.testPublicKey(SecretIT.java:59)
> ```
> 
> ```
> 10:38:21 [INFO ] [ERROR] Tests run: 12, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 3.693 s <<< FAILURE! - in com.walmartlabs.concord.it.server.TeamRbacIT
> 10:38:21 [INFO ] [ERROR] testSecretAccessLevels(com.walmartlabs.concord.it.server.TeamRbacIT)  Time elapsed: 0.33 s  <<< ERROR!
> 10:38:21 [INFO ] com.walmartlabs.concord.ApiException: Server response: The current user doesn't have the necessary access level (OWNER) to the secret: secret_1599057498237_7a008a
> 10:38:21 [INFO ] 	at com.walmartlabs.concord.it.server.TeamRbacIT.testSecretAccessLevels(TeamRbacIT.java:728)
> ```

that's strange. none of the ITs were failing on my local a week ago when I requested review on this. let me re-check

**Comment by d0s056f** [comment]
finally!

**Comment by ibodrov** [comment]
Changed it a bit to avoid passing the project's name from the UI.

**Review by ibodrov — APPROVED** [review]

**Pull request merged**
